xbmc/addons/kodi-addon-dev-kit/include/kodi/AddonBase.h

```diff
@@ -0,0 +1,439 @@
+{
+  // String with full path where add-on is installed (without his name on end)
+  // Set from Kodi!
+  const char* libBasePath;
+
+  // Pointer of first created instance, used in case this add-on goes with single way
+  // Set from Kodi!
+  KODI_HANDLE firstKodiInstance;
+
+  // Pointer to master base class inside add-on
+  // Set from addon header!
+  kodi::addon::CAddonBase* addonBase;
+
+  // Pointer to a instance used on single way (together with this class)
+  // Set from addon header!
+  kodi::addon::IAddonInstance* globalSingleInstance;
+
+  // Callback function tables from addon to Kodi
+  // Set from Kodi!
+  AddonToKodiFuncTable_Addon toKodi;
+
+  // Function tables from Kodi to addon
+  // Set from addon header!
+  KodiToAddonFuncTable_Addon toAddon;
+} AddonGlobalInterface;
+
+} /* extern "C" */
+
+//==============================================================================
+namespace kodi {
+namespace addon {
+/*
+ * Internal class to control various instance types with general parts defined
+ * here.
+ *
+ * Mainly is this currently used to identify requested instance types.
+ *
+ * @note This class is not need to know during add-on development thats why
+ * commented with "*".
+ */
+class IAddonInstance
+{
+public:
+  IAddonInstance(ADDON_TYPE type) : m_type(type) { }
+  virtual ~IAddonInstance() { }
+
+  virtual ADDON_STATUS CreateInstance(int instanceType, std::string instanceID, KODI_HANDLE instance, KODI_HANDLE& addonInstance)
+  {
+    return ADDON_STATUS_NOT_IMPLEMENTED;
+  }
+
+  const ADDON_TYPE m_type;
+};
+} /* namespace addon */
+} /* namespace kodi */
+//------------------------------------------------------------------------------
+
+//==============================================================================
+namespace kodi {
+///
+class CSettingValue
+{
+public:
+  CSettingValue(const void *settingValue) : m_settingValue(settingValue) {}
+
+  bool empty() const { return (m_settingValue == nullptr) ? true : false; }
+  std::string GetString() const { return (char*)m_settingValue; }
+  int GetInt() const { return *(int*)m_settingValue; }
+  unsigned int GetUInt() const { return *(unsigned int*)m_settingValue; }
+  bool GetBoolean() const { return *(bool*)m_settingValue; }
+  float GetFloat() const { return *(float*)m_settingValue; }
+
+private:
+  const void *m_settingValue;
+};
+} /* namespace kodi */
+//------------------------------------------------------------------------------
+
+//==============================================================================
+namespace kodi {
+namespace addon {
+/// Add-on main instance class.
+/// @todo need bigger improvement and rework
+class CAddonBase
+{
+public:
+  CAddonBase()
+  {
+    CAddonBase::m_interface->toAddon.create_instance = ADDONBASE_CreateInstance;
+    CAddonBase::m_interface->toAddon.destroy_instance = ADDONBASE_DestroyInstance;
+    CAddonBase::m_interface->toAddon.set_setting = ADDONBASE_SetSetting;
+  }
+
+  virtual ~CAddonBase()
+  {
+  }
```

Comment:
Here and above is changed to the default, it works too.

The destructors are added there to destruct also addons own class. During tests I has seen that without them a destructor on addon itself was not called (don't know why).